pkg/cleaner/cleaner.go

```diff
@@ -2,131 +2,33 @@
-	c.mut.Unlock()
-}
+// Run starts the deleted app cleaner. Every pollSleepDuration, it compares the result of nsLister.List with the directories in the top level of gitHome on the local file system. On any error, it uses log messages to output a human readable description of what happened.
+func Run(gitHome string, nsLister kcl.NamespaceInterface) error {
 
-// localDirs returns all of the local directories immediately under gitHome that filter returns true for. filter will receive only the names of each of the top level directories (not their fully qualified paths), and should return true if it should be included in the output
-func localDirs(gitHome string, filter func(string) bool) ([]string, error) {
-	fileInfos, err := ioutil.ReadDir(gitHome)
+	watcher, err := nsLister.Watch(nil, nil, "")
 	if err != nil {
-		return nil, err
-	}
-	var ret []string
-	for _, fileInfo := range fileInfos {
-		nm := fileInfo.Name()
-		if len(nm) <= 0 || nm == "." || !fileInfo.IsDir() {
-			continue
-		}
-		if filter(nm) {
-			ret = append(ret, filepath.Join(gitHome, nm))
-		}
-	}
-	return ret, nil
-}
-
-// getDiff gets the directories that are not in namespaceList
-func getDiff(namespaceList []api.Namespace, dirs []string) []string {
-	var ret []string
-
-	// create a set of lowercase namespace names
-	namespacesSet := make(map[string]struct{})
-	for _, ns := range namespaceList {
-		lowerName := strings.ToLower(ns.Name)
-		namespacesSet[lowerName] = struct{}{}
-	}
-
-	// get dirs not in the namespaces set
-	for _, dir := range dirs {
-		lowerName := strings.ToLower(dir)
-		if _, ok := namespacesSet[lowerName]; !ok {
-			ret = append(ret, lowerName)
-		}
+		log.Printf("unable to get watch events (%s)", err)
 	}
-
-	return ret
-}
-
-func stripSuffixes(strs []string, suffix string) []string {
-	ret := make([]string, len(strs))
-	for i, str := range strs {
-		idx := strings.LastIndex(str, suffix)
-		if idx >= 0 {
-			ret[i] = str[:idx]
-		} else {
-			ret[i] = str
-		}
-	}
-	return ret
-}
-
-func dirHasGitSuffix(dir string) bool {
-	return strings.HasSuffix(dir, dotGitSuffix)
-}
-
-// Run starts the deleted app cleaner. Every pollSleepDuration, it compares the result of nsLister.List with the directories in the top level of gitHome on the local file system. On any error, it uses log messages to output a human readable description of what happened.
-func (c Ref) Run(gitHome string, nsLister k8s.NamespaceLister, ref Ref, pollSleepDuration time.Duration) error {
 	for {
-		nsList, err := nsLister.List(labels.Everything(), fields.Everything())
-		if err != nil {
-			log.Printf("Cleaner error listing namespaces (%s)", err)
-			continue
-		} else {
-			lst := make([]string, len(nsList.Items))
-			for i, ns := range nsList.Items {
-				lst[i] = strings.ToLower(ns.Name)
+		event := <-watcher.ResultChan()
+		namespace := event.Object.(*api.Namespace)
+		if event.Type == "DELETED" {
+			if err := os.RemoveAll(namespace.ObjectMeta.Name + dotGitSuffix); err != nil {
```

Comment:
Builder state becomes inconsistent for a deleted App's git push failure is fine cause any ways the git push will fail while doing config, build or push hook. The only solution to make it cleaner is to have a hook from controller to check if the APP exists and then proceed accordingly. This can be handled later.